backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/sort/nested/impl/NestedNumericFieldComparatorSource.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.types.sort.nested.impl;
+
+import org.hibernate.search.backend.lucene.types.lowlevel.impl.LuceneNumericDomain;
+
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.Query;
+
+public class NestedNumericFieldComparatorSource<E extends Number> extends NestedFieldComparatorSource {
+
+	private final E missingValue;
+	private final LuceneNumericDomain<E> numericDomain;
+
+	private LuceneNumericDomain.NumericNestedFieldComparator<E> fieldComparator;
+
+	public NestedNumericFieldComparatorSource(String nestedDocumentPath, LuceneNumericDomain<E> numericDomain, E missingValue) {
+		super( nestedDocumentPath );
+		this.numericDomain = numericDomain;
+		this.missingValue = missingValue;
+	}
+
+	@Override
+	public FieldComparator<?> newComparator(String fieldname, int numHits, int sortPos, boolean reversed) {
+		fieldComparator = numericDomain.createNestedFieldComparator( fieldname, numHits, missingValue );
+
+		// workaround since we cannot extend a FieldComparator
+		return fieldComparator.getComparator();
```

Comment:
1. `new FieldComparator.TermOrdValComparator( numHits, fieldname, sortMissingLast )` is invoked  during LuceneCollectors is being built.
2. `setOriginalParentQuery(Query luceneQuery)` that allows to have a `nestedDocsProvider` is invoked at the very beginning of `LuceneCollectors#collect`.
3. `getSortedDocValues(LeafReaderContext context, String field)` is invoked always on `LuceneCollectors#collect` a bit later than the previous one.

So the field `nestedDocProvider` is not set yet when `newComparator` method is called.

To handle that we should postpone the creation of the comparator (and the `TopFieldCollector`) from `LuceneCollectorsBuilder#build` to `LuceneCollectors#collect`, when we finally have the `Query luceneQuery`.

Reply:
I see.

Maybe you could move the call to `org.hibernate.search.backend.lucene.types.sort.comparatorsource.impl.LuceneFieldComparatorSource#setOriginalParentQuery` out of `org.hibernate.search.backend.lucene.search.extraction.impl.LuceneCollectors#collect` and to `org.hibernate.search.backend.lucene.search.query.impl.LuceneSearchQueryBuilder#build` instead? `definitiveLuceneQuery` is the exact same query that is currently passed to `LuceneCollectors#collect`.

I think this would allow you to get rid of the `nestedFieldSorts` contructor parameter in `LuceneSearchQueryRequestContext`, too, as the sorts will be completely initialized straight out of the query builder.

Reply:
Yes. You're right the query builder time is perfect. Done